docs/design/features/bridging-constraints.md

```diff
@@ -0,0 +1,110 @@
+// Not expected to compile as C# in this shape, but imagine the straightforward lowering to IL.
+//
+interface IInterface { }
+
+struct Constrained<T> where T : IInterface { }
+
+static void Do<T>()
+{
+    if (typeof(T).IsAssignableFrom(typeof(IInterface))
+    {
+        Constrained<T> someLocal = default;
+        try
+        {
+            someLocal.ToString();
+        }
+        catch (GenericException<Constrained<T>>) { }
+    }
+}
+```
+
+The requirement of never having an invalid type instantiation within the runtime type system means that if we're compiling `Do<int>` (where we cannot allow `Constrained<int>` to exist), the code generator cannot ask questions about all locals (IL locals are scoped to the method, not to a basic block), exception handling regions, or any IL within the `IsAssignableFrom` check. While it may be reasonable to expect the `IsAssignableFrom` check gets optimized out, it would likely not happen for unoptimized code. Locals and EH regions pose additional programs.
+
+It would greatly simplify the codegen impact of this if we were to limit the amount of places within a method body that are allowed to break the constraints. The least impactful and potentially sufficiently powerful change would be allowing to break constraints on `call` instruction.
+
+The proposal is to:
+
+* Stop doing the verification-level constraint validation when compiling `call` (i.e. the check on uninstantiated forms).
+* Keep constraint validation of the instantiated form.
+* If the instantiated form of the call target doesn't satisfy constraint, replace the call with a call to a throw helper (we can likely reuse infrastructure around `CORINFO_ACCESS_ILLEGAL`).
+* If the constraint validation requires a runtime check due to shared code, delay validation to run time (this can potentially be optimized out if the potential constraint violation is already in a guarded block).
+
+### Alternative option: allow loading constraint-failing MethodTables
+
+Limiting the constraint-violating code to `call` instruction is required to simplify code generation. Allowing `MethodTable`s that don't satisfy their own constraints to load sounds like trouble. We could potentially allow them to load and then try to prevent them from leaking out. Calling this out for completeness, I don't know if it's a good idea.
+
+### Potential shortcut: Stop validating constraints
+
+One more avenue could be to drop all constraint checks within the runtime and make constraints source compiler's concern only. We already have a constraint (`unmanaged` in C#) that is only known to the source compiler and not enforced at run time. This would likely require some thoughts around failure modes when invalid IL within a method body IL is encountered. This document is not going to explore this direction further since the assumption is that we do not want to drop constraint checks from the runtime, we just want to delay them.
```

Comment:
Something that came to me right as I went to bed: If we get the ability to check generic constraints (like `unmanaged`) using runtime APIs, could that also open the door for new kinds of constraints? As a concrete example (pun half-intended) there's been places where it'd be real nice to constrain a type parameter to only concrete types (one such case has to do with `static abstract` members in interfaces, another is when `T` is something that would get instantiated with reflection such as in a DI container or the like).

The idea here is that for such a new constraint the compiler would emit a call to the appropriate helper method which enforces the appropriate behavior at runtime without needing a new IL encoding of said constraint.